rfcs/20250520-async-threadpool-support/README.md

```diff
@@ -0,0 +1,157 @@
+  primitive execution. As a result, multiple primitives will be
+  submitted from a threadpool to that same threadpool, so parallelism
+  will always be nested.
+
+We have some properties that we can leverage on oneDNN side:
+- Primitive and memory object lifetimes will be maintained by thunk
+  runtime while execution happens: no need to reference count oneDNN
+  objects to extend their lifetime.
+- consecutive `parallel_for` calls happen in order (similar to sycl
+  in\_order queues). We can leverage that to implicitly synchronize
+  multiple parallel regions.
+
+## API changes
+
+### Waiting on threadpool tasks completion
+
+Currently the external threadpool interface that users need to
+implement already exposes flags, to inform oneDNN implementation of
+the threadpool implementation properties.  The only flag we expose as
+of now is the `ASYNCHRONOUS` flag, which when set, implies that the
+threadpool iface implementation has a non-blocking implementation of
+the parallel method.
+
+However, until now, when the threadpool iface implementation was
+ASYNCHRONOUS, we used to explicit synchronize inside oneDNN library,
+and we plan on removing that synchronization. With that change, we
+need to add a wait() method to the threadpool iface, so that we can
+properly handle waiting through threadpool in stream:wait() method.
+
+Additions to threadpool_iface.hpp:
+```c++
+struct threadpool_iface {
+     // Does nothing if SYNCHRONOUS, waits for all jobs for ASYNCHRONOUS 
+     virtual void wait() = 0;
+}
+```
+
+### Dependency tracking
+
+Technically, we need to expose events or some dependency tracking
+mecanism to express dependencies between different `parallel_for`
+calls. However, because thunk runtime has implicit synchronization in
+a threadpool, we can leverage that.
+
+Here there are two options:
+- (recommended) no API extension is needed. We need to document that the ASYNCHRONOUS flag implies this property
+- we expose a new `IN_ORDER` flag, and we would require it to be set when ASYNCHORNOUS is set.
+
+## Internal changes
+
+The main idea to support asynchronous execution in oneDNN
+implementation will be to tie the lifetime of temporary objects during
+excecution to the closure of the std::function object (often a lambda)
+that will be submited to the threapool. We then can transfert the
+ownership of this std::function to the threadpool implementation, that
+will maintain the std::function object alive while it is beeing
+executed.
+
+### Managing std::function lifetime
+
+Here the change can be local to src/common/dnnl_thread.hpp, where we need to:
+- when redirecting internal `parallel*` to threadpool implemenation
+  `parallel_for`, we need to stop capturing references to variables in
+  lambda, but copy those variables. In general, those are just `dim_t`
+  variables and should have no additional overhead.
+- we should move ownership of these function objects through std::move
```

Comment:
If there are two lambdas and the 1st lambda calls the 2nd lambda and the 1st lambda is submitted to `parallel_for` using `std::move` then how will the 2nd lambda be handled?

Reply:
The thing I was thinking about isn't a valid concern so we should be fine.